pkg/filesystem/local_directory_windows.go

```diff
@@ -439,50 +439,60 @@ func (d *localDirectory) Mknod(name path.Component, perm os.FileMode, deviceNumb
 	return status.Error(codes.Unimplemented, "Creation of device nodes is not supported on Windows")
 }
 
+func clamp(value, max uint32) uint32 {
+	if value > max {
+		return max
+	}
+
+	return value
+}
+
 func readdirnames(handle windows.Handle) ([]string, error) {
-	outBufferSize := uint32(512)
-	outBuffer := make([]byte, outBufferSize)
-	firstIteration := true
+	outBufferSize := uint32(256)
+	names := make([]string, 0)
+
 	for {
-		err := windows.GetFileInformationByHandleEx(handle, windows.FileFullDirectoryInfo,
-			&outBuffer[0], outBufferSize)
-		if err == nil {
-			break
-		}
-		if err.(syscall.Errno) == windows.ERROR_NO_MORE_FILES {
-			if firstIteration {
-				return []string{}, nil
+		// NB: In testing this was found to be about the largest buffer the API
+		// would fill. Any larger allocation would be unused and lead to
+		// resource exhaustion. We have not found good documentation for how to
+		// use this.
+		outBufferSize = clamp(outBufferSize*2, 62000)
+		outBuffer := make([]byte, outBufferSize)
+		err := windows.GetFileInformationByHandleEx(handle, windows.FileFullDirectoryInfo, &outBuffer[0], outBufferSize)
+		if err != nil {
+			if err.(syscall.Errno) == windows.ERROR_NO_MORE_FILES {
+				break
+			}
+			// NB: We have never seen `ERROR_MORE_DATA` during development,
+			// it seems it is never raised. But according to the docs is should be set
+			// and we should proceed with the happy path.
```

Comment:
That's because you've never attempted to read directories containing filenames that are more than 256 bytes long. Please remove this comment, as it is a valid case that we should handle.

Reply:
Aha, that makes sense. I will setup a test to verify that this works. Have you used this API before, or know of good documentation? I have mostly been tinkering with it manually to see what pops back out.


https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-getfileinformationbyhandleex
https://github.com/microsoft/Windows-classic-samples/blob/main/Samples/Win7Samples/winbase/io/extendedfileapis/ExtendedFileAPIs.cpp#L468

Reply:
I don't know anything about it. I don't use Windows myself. ;-)

Reply:
I'll need some more time to tinker with this.
I now have a manual test that hits the `ERROR_MORE_DATA` for a 250-character filename and a small initial buffer.
Will wire that up.

Have not been able to test against the long-file path API as it is hard to actually create such files. I'm not very familiar with the special APIs either.